File: dsa/algorithms/sorting_algos/merge_sort/merge_sort.py
def merge_sort(arr):
    if len(arr) <= 1:
        return arr
    left_arr = arr[:len(arr)//2]
    right_arr = arr[len(arr)//2:]

    merge_sort(left_arr)
    merge_sort(right_arr)

    i = 0
    j = 0
    k = 0
    while i < len(left_arr) and j < len(right_arr):
        if left_arr[i] < right_arr[j]:
            arr[k] = left_arr[i]
            i += 1
        else:
            arr[k] = right_arr[j]
            j += 1
        k += 1
    
    while i < len(left_arr):
        arr[k] = left_arr[i]
        k += 1
        i += 1

    while j < len(right_arr):
        arr[k] = right_arr[j]
        j += 1
        k += 1
    

arr = [2, 1, 2, 1, 2, 5, 4, 3, 2, 5, 3, 2, 1, 5, 7, 22, 72, 7, 2, 1, 6, 3, 5, 3, 5]
merge_sort(arr)
print(arr)

File: dsa/algorithms/sorting_algos/merge_sort/merge_sort.java
package merge_sort;

public class merge_sort {
    public static void main(String[] args) {
        int[] arr = {2, 4, 2, 5, 3, 5, 2, 5, 7, 8, 2, 4, 66, 1};
        int[] result = algo(arr);
        for (int i : result) {
            System.out.printf("%s ", i);
        }
    }

    static int[] algo(int[] arr) {
        if (arr.length <= 1) return arr;
        int[] left_arr = slice(arr, 0, arr.length / 2);
        int[] right_arr = slice(arr, arr.length / 2, arr.length);
        left_arr = algo(left_arr);
        right_arr = algo(right_arr);
        int i = 0;
        int j = 0;
        int k = 0;
        int[] result = new int[left_arr.length + right_arr.length];
        while (i < left_arr.length && j < right_arr.length) {
            if (left_arr[i] > right_arr[j]) {
                result[k] = right_arr[j];
                j++;
            } else {
                result[k] = left_arr[i];
                i++;
            }
            k++;
        }

        while (i < left_arr.length) {
            result[k] = left_arr[i];
            i++;
            k++;
        }
        while (j < right_arr.length) {
            result[k] = right_arr[j];
            j++;
            k++;
        }
        return result;
    }

    static int[] slice(int[] arr, int start, int end) {
        int[] arr2 = new int[end - start];
        int j = 0;
        for (int i = start; i < end; i++) {
            arr2[j] = arr[i];
            j += 1;
        } 
        return arr2;
    }
}
